File: client/java/src/avt/MsgDlg.java
package avt;

import java.util.Vector;
import javax.microedition.lcdui.Graphics;
import main.Canvas;

public final class MsgDlg extends Dialog {
   private Vector info;
   private String str = "";
   private Vector list = new Vector();
   private int index = 0;
   private int w;
   private int h;
   private int x;
   private int y;
   public boolean isWaiting = false;
   private int num = 0;
   private int colorIndex = 0;
   private static final int[] RAINBOW_COLORS = new int[]{0xFFFF0000, 0xFFFF7F00, 0xFFFFFF00, 0xFF00FF00, 0xFF0000FF, 0xFF4B0082, 0xFF9400D3, 0xFFFF0000, 0xFFFF7F00};
   public static FrameImage[] imgLoadRainbow = new FrameImage[9];
   private int size = 0;
   private int hText;
   private int hDu;
   private int indexLeft = 0;
   private int indexRight = 0;
   private int hCell;
   public static FrameImage imgLoad;
   private long timeDelay = -1L;
   private long limitTime;
   private long timeEnd;
   private int u = 0;
   private int spinAngle = 0;
   private static final int[] sinTable = new int[360];
   private static final int[] cosTable = new int[360];
   static {
      for (int i = 0; i < 360; i++) {
         sinTable[i] = (int)(Math.sin(Math.toRadians(i)) * 100);
         cosTable[i] = (int)(Math.cos(Math.toRadians(i)) * 100);
      }
   }

   public MsgDlg() {
      this.hText = AvMain.hBlack;
   }

   public final void setInfoC(String var1, Command var2, Vector var3) {
      if (ChatTextField.isShow) {
         ChatTextField.gI().closeChat();
      }

      this.hCell = MyScreen.hText;
      this.isWaiting = false;
      this.str = var1;
      super.center = var2;
      this.index = 0;
      this.list = var3;
      if (var3 != null) {
         Command var4 = (Command)var3.elementAt(this.index);
         super.center = var4;
         if (var4 != null) {
            super.center.action = var4.action;
            super.center.indexMenu = var4.indexMenu;
            super.center.pointer = var4.pointer;
         }

         this.u = 0;

         for(int var5 = 0; var5 < var3.size(); ++var5) {
            var2 = (Command)var3.elementAt(var5);
            if (Canvas.normalFont.getWidth(var2.caption) > this.u) {
               this.u = Canvas.normalFont.getWidth(var2.caption) + (Canvas.isKeyBoard ? this.w / 3 : 0);
            }
         }
      } else {
         this.timeEnd = System.currentTimeMillis() / 100L;
      }

      this.size = 0;
      if (this.list != null) {
         this.size = this.list.size();
      }

      this.num = 0;
      this.timeDelay = -1L;
      this.init();
      Canvas.currentDialog = Canvas.msgdlg;
   }

   public final void init() {
      this.w = Canvas.w - 80;
      if (Canvas.w < 200) {
         this.w = Canvas.w - 40;
         if (Canvas.w <= 128) {
            this.w = Canvas.w - 10;
         }
      }

      if (this.str.equals(T.pleaseWait)) {
         this.w = Canvas.hw;
      }

      this.info = Canvas.M.splitFontBStrInLineV(this.str, this.w - 16);
      this.h = this.info.size() * this.hText + 20;
      this.hDu = 0;
      if (super.center != null) {
         this.h += this.hCell + 15 * AvMain.hd;
         this.hDu += this.hCell + 15 * AvMain.hd;
      }

      if (this.h < this.hCell * 3 + (AvMain.hd - 1) * 15) {
         this.h = this.hCell * 3 + (AvMain.hd - 1) * 15;
      }

      this.x = Canvas.hw - this.w / 2;
      this.y = Canvas.hCan - Canvas.hTab - this.h - 10;
   }

   public final void setIsWaiting(boolean var1) {
      this.isWaiting = var1;
      this.h = this.info.size() * this.hText + 20;
      if (this.isWaiting) {
         this.h += 25 * AvMain.hd + 4;
         this.hDu += 25 * AvMain.hd + 4;
         this.colorIndex = 0;
      }

      int var2 = this.hCell * 3 + (AvMain.hd - 1) * 15;
      if (this.h < var2) {
         this.h = var2;
      }

      this.y = Canvas.hCan - Canvas.hTab - this.h - 10;
      this.limitTime = (long)Canvas.getSecond();
   }

   public final void paint(Graphics var1) {
      Canvas.resetTrans(var1);
      if (System.currentTimeMillis() / 100L - this.timeEnd >= 5L) {
         Canvas.paint.drawRectangle(var1, this.x, this.y, this.w, this.h, PaintPopup.color[0], PaintPopup.color[1], 0);
         if (super.center != null) {
            PaintPopup.fill(this.x + 1, this.y + this.h - (this.hCell + 15 * AvMain.hd - 4), this.w - 2, this.hCell, 15530985, var1);
         }

         if (this.isWaiting) {
            int cx = this.x + this.w / 2;
            int cy = this.y + 4 + (this.h - this.hDu) / 2 + this.info.size() * AvMain.hBlack / 2 + (this.h - (4 + (this.h - this.hDu) / 2 + this.info.size() * AvMain.hBlack / 2)) / 2;
            int radius = 7 * AvMain.hd;
            int dotRadius = 2 * AvMain.hd;
            int numDots = 8;
            int[] colors = new int[]{0xFFFF0000, 0xFFFFFF00, 0xFF00FF00, 0xFF00FFFF, 0xFF0000FF, 0xFFFF00FF, 0xFFFF0000, 0xFFFFFF00, 0xFF00FF00};
            for (int i = 0; i < numDots; i++) {
               int angle = this.spinAngle + (360 / numDots) * i;
               int dx = cx + (radius * cosTable[angle % 360]) / 100;
               int dy = cy - (radius * sinTable[angle % 360]) / 100;
               int color = colors[i % colors.length];
               var1.setColor(color);
               var1.fillArc(dx - dotRadius, dy - dotRadius, dotRadius * 2, dotRadius * 2, 0, 360);
            }
         }

         if (this.size > 0) {
            Command var2 = (Command)this.list.elementAt(this.index);
            Canvas.normalFont.drawString(var1, var2.caption, Canvas.hw, this.y + this.h - (this.hCell + 15 * AvMain.hd - 4) + this.hCell / 2 - AvMain.hNormal / 2, 2);
            if (this.size > 1) {
               Canvas.paint.drawHighlightedArea(var1, Canvas.hw - this.u / 2 - 11, (Canvas.stypeInt != 2 ? AvMain.hNormal / 2 : 0) + this.y + this.h - (this.hCell + 15 * AvMain.hd - 4) + MyScreen.hText / 2 + 1 + (Canvas.stypeInt == 1 ? -7 : 0) + (Canvas.stypeInt == 0 ? -3 : 0), 17 + this.u, this.indexLeft / 3, this.indexRight / 3);
            }
         } else if (super.center != null) {
            Canvas.normalFont.drawString(var1, super.center.caption, Canvas.hw, this.y + this.h - (this.hCell + 15 * AvMain.hd - 4) + this.hCell / 2 - AvMain.hNormal / 2, 2);
         }

         for(int var3 = 0; var3 < this.info.size(); ++var3) {
            Canvas.M.drawString(var1, (String)this.info.elementAt(var3), Canvas.hw, this.y + 4 + (this.h - this.hDu) / 2 - this.info.size() * AvMain.hBlack / 2 + var3 * AvMain.hBlack, 2);
         }
      }

   }

   public final void commandTab(int var1, int var2) {
      switch (var1) {
         case -2:
            MapScr.gI().doExitGame();
            return;
         case -1:
            this.isWaiting = false;
            Canvas.currentDialog = null;
            return;
         default:
            Canvas.currentMyScreen.commandTab(var1, var2);
      }
   }

   private void setIndex(int var1) {
      if (this.size > 0) {
         this.index += var1;
         if (this.index < 0) {
            this.index = this.size - 1;
         }

         if (this.index >= this.size) {
            this.index = 0;
         }

         Command var2 = (Command)this.list.elementAt(this.index);
         super.center = var2;
      }

   }

   public final void updateKey() {
      int var2;
      if (this.isWaiting) {
         ++this.num;
         if (this.num >= 8) {
            this.num = 0;
         }

         ++this.colorIndex;
         if (this.colorIndex >= 8) {
            this.colorIndex = 0;
         }

         this.spinAngle = (this.spinAngle + 15) % 360;

         if ((long)Canvas.getSecond() - this.limitTime > 30L) {
            String var1 = "";

            for(var2 = 0; var2 < this.info.size(); ++var2) {
               var1 = var1 + (String)this.info.elementAt(var2) + " ";
            }

            Canvas.startOK(var1, -2, (AvMain)null);
         }
      }

      if (this.timeDelay != -1L && System.currentTimeMillis() / 100L - this.timeDelay > 0L) {
         Canvas.keyPressed[5] = true;
      }

      if (this.indexLeft > 0) {
         --this.indexLeft;
      }

      if (this.indexRight > 0) {
         --this.indexRight;
      }

      if (Canvas.isKeyPressed(4)) {
         this.setIndex(-1);
         this.indexLeft = 5;
      } else if (Canvas.isKeyPressed(6)) {
         this.setIndex(1);
         this.indexRight = 5;
      }

      if (Canvas.isPointerRelease) {
         label85: {
            int var3 = 0;
            if (this.list != null && this.list.size() > 0) {
               Command var4 = (Command)this.list.elementAt(this.index);
               var3 = Canvas.normalFont.getWidth(var4.caption) + 20;
            } else if (super.center != null) {
               var3 = Canvas.normalFont.getWidth(super.center.caption) + 20;
            }

            var3 *= AvMain.hd;
            if (super.center != null && Canvas.isPointer(Canvas.hw - var3 / 2, this.y + this.h - (this.hCell + 18 * AvMain.hd - 4), var3, this.hCell)) {
               Canvas.endDlg();
               this.perform(super.center);
            } else {
               if (!Canvas.isPointer(this.x + 1, this.y + this.h - (this.hCell + 18 * AvMain.hd - 4), this.w - 2, this.hCell)) {
                  break label85;
               }

               if ((var2 = Canvas.hw - Canvas.px) > var3 / 2) {
                  this.setIndex(-1);
                  this.indexLeft = 5;
               } else if (var2 < -var3 / 2) {
                  this.setIndex(1);
                  this.indexRight = 5;
               }
            }

            Canvas.isPointerRelease = false;
         }
      }

      super.updateKey();
   }
}

File: client/unity/Assets/Scripts/Assembly-CSharp/MsgDlg.cs
public class MsgDlg : Dialog
{
	protected MyVector info;

	private string str = string.Empty;

	public MyVector list = new MyVector();

	public int index;

	public int w;

	public int h;

	public int x;

	public int y;

	public int yWait;

	public bool isWaiting;

	private int num;

	private int size;

	private int hText;

	private int hDu;

	private new int indexLeft;

	private new int indexRight;

	public static int hCell;

	public static FrameImage imgLoad;

	public static Image imgLoadOn;

	private long timeDelay = -1L;

	private long limitTime;

	private long timeEnd;

	public static bool isBack = true;

	public long timeOpen;

	public static sbyte[] fr = new sbyte[10] { 0, 0, 0, 0, 0, 1, 1, 1, 1, 1 };

	public int xCow;

	private int wStr;

	private int frame;

	private int direct;

	public MsgDlg()
	{
		hText = AvMain.hBlack;
	}

	public override void show()
	{
	}

	public void setInfoLR(string info, Command lef, Command righ)
	{
		isWaiting = false;
		Canvas.currentDialog = null;
		w = Canvas.w - 100 * AvMain.hd;
		if (w > 300 * AvMain.hd)
		{
			w = 300 * AvMain.hd;
		}
		str = info;
		index = 0;
		timeDelay = -1L;
		left = lef;
		right = righ;
		center = null;
		init();
		Canvas.currentDialog = this;
	}

	public void setInfoLCR(string info, Command lef, Command cente, Command righ)
	{
		Canvas.currentDialog = null;
		isWaiting = false;
		w = Canvas.w - 60 * AvMain.hd;
		if (w > 350 * AvMain.hd)
		{
			w = 350 * AvMain.hd;
		}
		str = info;
		index = 0;
		timeDelay = -1L;
		left = lef;
		right = righ;
		center = cente;
		init();
		Canvas.currentDialog = this;
	}

	public void setInfoC(string info, Command cente)
	{
		isWaiting = false;
		Canvas.currentDialog = null;
		w = Canvas.w - 100 * AvMain.hd;
		if (w > 300 * AvMain.hd)
		{
			w = 300 * AvMain.hd;
		}
		if (cente == null && w > 500)
		{
			w = 500;
		}
		str = info;
		index = 0;
		timeDelay = -1L;
		left = null;
		right = null;
		center = null;
		center = cente;
		init();
		Canvas.currentDialog = this;
	}

	public void setDelay(int limit)
	{
		limitTime = limit;
		timeDelay = Canvas.getTick() / 100;
	}

	public void init()
	{
		if (Canvas.currentDialog == null && left == null && center == null && right == null)
		{
			timeOpen = Canvas.getTick() / 100;
			list.removeAllElements();
		}
		if (Canvas.load == 0 && onMainMenu.isOngame)
		{
			Canvas.load = -1;
		}
		isBack = true;
		if (str.Equals(T.pleaseWait))
		{
			w = Canvas.hw;
		}
		info = Canvas.tempFont.splitFontBStrInLineV(str, w - w / 3);
		h = AvMain.hCmd / 2 + 10 + info.size() * Canvas.tempFont.getHeight() + 30 * AvMain.hd;
		hDu = 0;
		x = Canvas.hw - w / 2;
		y = Canvas.hCan / 2 - h / 2;
		if (onMainMenu.isOngame)
		{
			if (center != null)
			{
				center.x = -1;
			}
			if (left != null)
			{
				left.x = -1;
			}
			if (right != null)
			{
				right.x = -1;
			}
		}
		else
		{
			if (center == null && left != null && right != null)
			{
				left.x = x + w / 2 - MyScreen.wTab / 2 - 10 * AvMain.hd;
				left.y = (right.y = y + h - AvMain.hCmd / 2);
				right.x = x + w / 2 + MyScreen.wTab / 2 + 10 * AvMain.hd;
			}
			if (left != null && center != null && right != null)
			{
				left.x = x + w / 2 - MyScreen.wTab - 20;
				left.y = y + h - AvMain.hCmd / 2;
				right.x = x + w / 2 + MyScreen.wTab + 20;
				right.y = y + h - AvMain.hCmd / 2;
				center.x = Canvas.hw;
				center.y = y + h - AvMain.hCmd / 2;
			}
			if (left == null && right == null && center != null)
			{
				center.y = y + h - AvMain.hCmd / 2;
				center.x = Canvas.hw;
			}
		}
		if (onMainMenu.isOngame)
		{
			y = Canvas.hCan - 36 * AvMain.hd - h - 10;
		}
	}

	public void setIsWaiting(bool isa)
	{
		isWaiting = isa;
		timeEnd = Canvas.getSecond();
	}

	public override void paint(MyGraphics g)
	{
		Canvas.resetTrans(g);
		if (isWaiting && Canvas.getTick() / 100 - timeOpen < 4)
		{
			return;
		}
		if (isBack && !onMainMenu.isOngame)
		{
			Canvas.paint.paintTransBack(g);
		}
		if (onMainMenu.isOngame)
		{
			g.drawImageScale(imgLoadOn, x, y, w, h, 0);
		}
		else
		{
			Canvas.paint.paintPopupBack(g, x, y, w, h, -1, false);
		}
		int num = y + 4 * AvMain.hd + (h - 4 * AvMain.hd) / 2 - Canvas.tempFont.getHeight() * info.size() / 2 - 2 * AvMain.hd;
		if (left != null || center != null || right != null)
		{
			num = y + 4 * AvMain.hd + (h - 4 * AvMain.hd - AvMain.hCmd / 2) / 2 - Canvas.tempFont.getHeight() * info.size() / 2 - 2 * AvMain.hd;
		}
		if (isWaiting)
		{
			imgLoad.drawFrame(this.num / 2, x + w / 2 - Canvas.tempFont.getWidth((string)info.elementAt(0)) / 2 - 30 * AvMain.hd, y + h / 2, 0, 3, g);
		}
		for (int i = 0; i < info.size(); i++)
		{
			if (onMainMenu.isOngame)
			{
				Canvas.menuFont.drawString(g, (string)info.elementAt(i), Canvas.hw, y + h / 2 - Canvas.menuFont.getHeight() * info.size() / 2 + i * Canvas.tempFont.getHeight(), 2);
			}
			else
			{
				Canvas.tempFont.drawString(g, (string)info.elementAt(i), Canvas.hw, num + i * Canvas.tempFont.getHeight() - 3, 2);
			}
		}
		if (onMainMenu.isOngame)
		{
			Canvas.paint.paintTabSoft(g);
			Canvas.paint.paintCmdBar(g, left, center, right);
		}
		else
		{
			base.paint(g);
		}
	}

	public override void commandTab(int index)
	{
		switch (index)
		{
		case -2:
			MapScr.gI().doExitGame();
			break;
		case -1:
			isWaiting = false;
			Canvas.currentDialog = null;
			break;
		default:
			Canvas.currentMyScreen.commandTab(index);
			break;
		}
	}

	private void setIndex(int ina)
	{
		if (size > 0)
		{
			index += ina;
			if (index < 0)
			{
				index = size - 1;
			}
			if (index >= size)
			{
				index = 0;
			}
			Command command = (Command)list.elementAt(index);
			center.action = command.action;
			center.indexMenu = command.indexMenu;
		}
	}

	public override void updateKey()
	{
		if (onMainMenu.isOngame)
		{
			Canvas.paint.updateKeyOn(left, center, right);
		}
		else
		{
			base.updateKey();
		}
		if (isWaiting)
		{
			num++;
			if (num >= 16)
			{
				num = 0;
			}
			if (Canvas.getSecond() - timeEnd > 30)
			{
				isWaiting = false;
				string text = string.Empty;
				for (int i = 0; i < info.size(); i++)
				{
					text = text + (string)info.elementAt(i) + " ";
				}
				Canvas.startOKDlg(T.doYouWantExit2, new GlobalLogicHandler.IActionDisconnect());
			}
		}
		if (timeDelay == -1 || Canvas.getTick() / 100 - timeDelay > limitTime)
		{
		}
		if (indexLeft > 0)
		{
			indexLeft--;
		}
		if (indexRight > 0)
		{
			indexRight--;
		}
		if (Canvas.isPointerClick)
		{
			if (!isWaiting && size > 0)
			{
				if (Canvas.isPointer(Canvas.hw - wStr / 2 - 11 - hCell, ((AvMain.hd != 2) ? (Canvas.normalFont.getHeight() / 2) : 0) + y + h - (hCell + 15 * AvMain.hd - 4) + hCell / 2 + 1 + ((AvMain.hd == 1) ? (-7) : 0) + ((AvMain.hd == 0) ? (-3) : 0) - hCell / 2, hCell * 2, hCell))
				{
					setIndex(-1);
					indexLeft = 5;
					Canvas.isPointerRelease = false;
				}
				else if (Canvas.isPointer(Canvas.hw - wStr / 2 - 11 - hCell + wStr + 20, ((AvMain.hd != 2) ? (Canvas.normalFont.getHeight() / 2) : 0) + y + h - (hCell + 15 * AvMain.hd - 4) + hCell / 2 + 1 + ((AvMain.hd == 1) ? (-7) : 0) + ((AvMain.hd == 0) ? (-3) : 0) - hCell / 2, hCell * 2, hCell))
				{
					setIndex(1);
					indexRight = 5;
					Canvas.isPointerRelease = false;
				}
			}
			if (!isWaiting && Canvas.isPointer(Canvas.hw - hCell, ((AvMain.hd != 2) ? (Canvas.normalFont.getHeight() / 2) : 0) + y + h - (hCell + 15 * AvMain.hd - 4) + hCell / 2 + 1 + ((AvMain.hd == 1) ? (-7) : 0) + ((AvMain.hd == 0) ? (-3) : 0) - hCell / 2, hCell * 2, hCell))
			{
				Canvas.endDlg();
				perform(center);
				Canvas.isPointerRelease = false;
				Canvas.paint.clickSound();
			}
		}
		base.updateKey();
	}
}
